File: mobile/lib/shared/offline/sync_manager.dart
import 'dart:convert';

import 'package:sqflite/sqflite.dart';
import 'package:path/path.dart';

import 'pending_upload.dart';

/// Offline upload queue (mobile.md offline-first rules).
///
/// All write operations (create capture_session, create passport, upload
/// frames) are enqueued here with a client-generated `idempotency_key`, then
/// drained when connectivity is restored.
///
/// This is an interface so it can be backed by SQLite on device
/// ([SqfliteSyncManager]) or by memory on web/tests ([InMemorySyncManager]) —
/// sqflite has no web implementation.
abstract class SyncManager {
  /// After this many failed attempts an item is surfaced to the user as an
  /// error rather than retried (mobile.md: max retry 5).
  static const int maxRetries = 5;

  Future<void> enqueue({
    required String idempotencyKey,
    required String operationType,
    required Map<String, dynamic> payload,
  });

  Future<List<PendingUpload>> getPendingUploads({int limit = 20});

  Future<void> markCompleted(int id);

  Future<void> incrementRetry(int id);

  Future<int> pendingCount();
}

/// SQLite-backed queue for Android/iOS/desktop.
class SqfliteSyncManager implements SyncManager {
  SqfliteSyncManager({Database? db}) : _db = db;

  Database? _db;

  Future<Database> get _database async => _db ??= await _initDb();

  Future<Database> _initDb() async {
    final path = join(await getDatabasesPath(), 'zmetrics_sync.db');
    return openDatabase(
      path,
      version: 1,
      onCreate: (db, version) async {
        await db.execute('''
          CREATE TABLE pending_uploads (
            id INTEGER PRIMARY KEY AUTOINCREMENT,
            idempotency_key TEXT NOT NULL UNIQUE,
            operation_type TEXT NOT NULL,
            payload_json TEXT NOT NULL,
            retry_count INTEGER NOT NULL DEFAULT 0,
            created_at TEXT NOT NULL,
            last_attempt_at TEXT
          )
        ''');
      },
    );
  }

  @override
  Future<void> enqueue({
    required String idempotencyKey,
    required String operationType,
    required Map<String, dynamic> payload,
  }) async {
    final db = await _database;
    await db.insert(
      'pending_uploads',
      {
        'idempotency_key': idempotencyKey,
        'operation_type': operationType,
        'payload_json': jsonEncode(payload),
        'retry_count': 0,
        'created_at': DateTime.now().toIso8601String(),
      },
      // Re-enqueuing the same idempotency key is a no-op (dedupe).
      conflictAlgorithm: ConflictAlgorithm.ignore,
    );
  }

  @override
  Future<List<PendingUpload>> getPendingUploads({int limit = 20}) async {
    final db = await _database;
    final rows = await db.query(
      'pending_uploads',
      orderBy: 'created_at ASC',
      limit: limit,
    );
    return rows.map(PendingUpload.fromRow).toList();
  }

  @override
  Future<void> markCompleted(int id) async {
    final db = await _database;
    await db.delete('pending_uploads', where: 'id = ?', whereArgs: [id]);
  }

  @override
  Future<void> incrementRetry(int id) async {
    final db = await _database;
    await db.rawUpdate(
      'UPDATE pending_uploads SET retry_count = retry_count + 1, '
      'last_attempt_at = ? WHERE id = ?',
      [DateTime.now().toIso8601String(), id],
    );
  }

  @override
  Future<int> pendingCount() async {
    final db = await _database;
    final result =
        await db.rawQuery('SELECT COUNT(*) as count FROM pending_uploads');
    return Sqflite.firstIntValue(result) ?? 0;
  }
}

/// In-memory queue for Flutter Web (no sqflite) and unit tests.
class InMemorySyncManager implements SyncManager {
  final List<PendingUpload> _items = [];
  int _seq = 0;

  @override
  Future<void> enqueue({
    required String idempotencyKey,
    required String operationType,
    required Map<String, dynamic> payload,
  }) async {
    // Dedupe on idempotency key, matching the SQLite UNIQUE constraint.
    if (_items.any((e) => e.idempotencyKey == idempotencyKey)) return;
    _items.add(PendingUpload(
      id: ++_seq,
      idempotencyKey: idempotencyKey,
      operationType: operationType,
      payload: payload,
      retryCount: 0,
      createdAt: DateTime.now(),
    ));
  }

  @override
  Future<List<PendingUpload>> getPendingUploads({int limit = 20}) async {
    final sorted = [..._items]
      ..sort((a, b) => a.createdAt.compareTo(b.createdAt));
    return sorted.take(limit).toList();
  }

  @override
  Future<void> markCompleted(int id) async {
    _items.removeWhere((e) => e.id == id);
  }

  @override
  Future<void> incrementRetry(int id) async {
    final idx = _items.indexWhere((e) => e.id == id);
    if (idx == -1) return;
    final cur = _items[idx];
    _items[idx] = PendingUpload(
      id: cur.id,
      idempotencyKey: cur.idempotencyKey,
      operationType: cur.operationType,
      payload: cur.payload,
      retryCount: cur.retryCount + 1,
      createdAt: cur.createdAt,
      lastAttemptAt: DateTime.now(),
    );
  }

  @override
  Future<int> pendingCount() async => _items.length;
}

File: desktop/zmetrics_desktop/offline/sync_manager.py
"""SQLite-backed offline queue for write operations.

Parity with the former Flutter ``SyncManager``: pending operations live in the
``pending_uploads`` table, each with a client-generated ``idempotency_key`` (UUID v4).
The queue is drained when connectivity is available; after ``MAX_ATTEMPTS`` failures an
item is surfaced to the user as an error rather than retried forever.
"""
from __future__ import annotations

import json
import sqlite3
import uuid
from dataclasses import dataclass
from datetime import datetime, timezone
from pathlib import Path

MAX_ATTEMPTS = 5

_SCHEMA = """
CREATE TABLE IF NOT EXISTS pending_uploads (
    idempotency_key TEXT PRIMARY KEY,
    kind            TEXT NOT NULL,
    payload         TEXT NOT NULL,
    attempts        INTEGER NOT NULL DEFAULT 0,
    last_error      TEXT,
    created_at      TEXT NOT NULL
);
"""


@dataclass(frozen=True)
class PendingUpload:
    idempotency_key: str
    kind: str
    payload: dict
    attempts: int
    last_error: str | None
    created_at: str


class SyncManager:
    """Minimal durable queue. Not thread-safe; use one instance per worker thread."""

    def __init__(self, db_path: str | Path = ":memory:") -> None:
        if db_path != ":memory:":
            Path(db_path).parent.mkdir(parents=True, exist_ok=True)
        self._conn = sqlite3.connect(str(db_path))
        self._conn.row_factory = sqlite3.Row
        self._conn.executescript(_SCHEMA)
        self._conn.commit()

    def close(self) -> None:
        self._conn.close()

    def enqueue(self, kind: str, payload: dict) -> str:
        """Queue an operation; returns its idempotency_key (UUID v4)."""
        key = str(uuid.uuid4())
        self._conn.execute(
            "INSERT INTO pending_uploads (idempotency_key, kind, payload, created_at) "
            "VALUES (?, ?, ?, ?)",
            (key, kind, json.dumps(payload), datetime.now(timezone.utc).isoformat()),
        )
        self._conn.commit()
        return key

    def pending(self) -> list[PendingUpload]:
        """Items still eligible for sending (attempts < MAX_ATTEMPTS), oldest first."""
        rows = self._conn.execute(
            "SELECT * FROM pending_uploads WHERE attempts < ? ORDER BY created_at ASC",
            (MAX_ATTEMPTS,),
        ).fetchall()
        return [self._row_to_item(r) for r in rows]

    def failed(self) -> list[PendingUpload]:
        """Items exhausted (attempts >= MAX_ATTEMPTS) — surface these to the user."""
        rows = self._conn.execute(
            "SELECT * FROM pending_uploads WHERE attempts >= ? ORDER BY created_at ASC",
            (MAX_ATTEMPTS,),
        ).fetchall()
        return [self._row_to_item(r) for r in rows]

    def mark_failure(self, key: str, error: str) -> None:
        self._conn.execute(
            "UPDATE pending_uploads SET attempts = attempts + 1, last_error = ? "
            "WHERE idempotency_key = ?",
            (error, key),
        )
        self._conn.commit()

    def remove(self, key: str) -> None:
        """Remove an item after the server confirmed it (success)."""
        self._conn.execute(
            "DELETE FROM pending_uploads WHERE idempotency_key = ?", (key,)
        )
        self._conn.commit()

    @staticmethod
    def _row_to_item(row: sqlite3.Row) -> PendingUpload:
        return PendingUpload(
            idempotency_key=row["idempotency_key"],
            kind=row["kind"],
            payload=json.loads(row["payload"]),
            attempts=row["attempts"],
            last_error=row["last_error"],
            created_at=row["created_at"],
        )
